### PR TITLE
[Feature] Handle auto-properties used in F# correctly

### PR DESCRIPTION
**Pull request by robertcoltheart**
Per: https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/members/explicit-fields-the-val-keyword

Backing fields in F# auto-properties are suffixed with `@`.

Eg. This F# type:
```f#
[<CLIMutable>]
[<FixedLengthRecord>]
type SampleFSharpType = 
    { [<FieldFixedLength(8)>]
      [<FieldConverter(ConverterKind.Date, "ddMMyyyy")>]
      Field1: DateTime }
```
compiles down to (reversed C#):
```c#
    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
    internal DateTime Field1@;

    [FieldFixedLength(8)]
    [FieldConverter(ConverterKind.Date, "ddMMyyyy")]
    [CompilationMapping(SourceConstructFlags.Field, 0)]
    public DateTime Field1
    {
      get
      {
        return this.Field1@;
      }
      set
      {
        this.Field1@ = value;
      }
    }
```

**Note:** I _really_ tried to make this work with compiling from F# within the test project using CodeDomProvider (the F# variant), but couldn't make it work. Hence the addition of a `FileHelpers.Tests.Fsharp` project which is referenced from the tests projects. Happy to see suggestions of how I can make this cleaner.

### CODE REVIEW

**Comment by robertcoltheart** [comment]
Not sure why some of the build is failing...

**Comment by robertcoltheart** [comment]
@mcavigelli are you able to take a look please?

**Comment by jonnii** [comment]
We've been bitten by this one too, would be great to get this in a new release.

**Comment by mcavigelli** [comment]
Hi Robert and Jonathan

Thank you for your contribution.
I will look into it tomorrow or the day after tomorrow.

Matthias

**Comment by mcavigelli** [comment]
Hi Robert

It code looks fine to me. I don't know of a simpler testing possibility.

But I have noticed that the F# project has some dependencies.
Would it be possible to remove some of those? I have not found out how to do that.

image

That would be an improvement. Is that possible?

Matthias

-----
On Appveyor it was working, now it also works on my machine.
I removed some "broken windows":
- travis (has mono support, with .NET Standard not that important any more IMHO)
- azure pipelines (no idea how that file got into the repository).

**Comment by robertcoltheart** [comment]
@mcavigelli I've tidied this up quite a bit and found a way to compile F# directly in C#, so no need for that F#-only project. However I needed to create a new test project that is not strongly-named, as the Fsharp compiler dll is also not strongly named, so I couldn't use it as part of the `FileHelpers.Tests` project. I had basically 2 alternatives:

1. Drop strong-naming for `FileHelpers.Tests` and then have to deal with accessing internal members via reflection, instead of using `InternalsVisibleTo`. This would be quite an invasive change to the tests.
2. Create a new test project that is not strongly-named and create the tests for F# there.

If we want to merge these 2 in the future to make it cleaner, we could do that, and just have to deal with reflection-type stuff then.

File: FileHelpers/Fields/FieldBase.cs [hunk 1]
```diff
@@ -387,6 +388,17 @@ internal static string AutoPropertyName(FieldInfo fi)
...
             }
 
+            if (fi.IsDefined(typeof(DebuggerBrowsableAttribute), false))
+            {
+                if (fi.Name.EndsWith("@") && !fi.IsPublic)
+                {
+                    var name = fi.Name.Substring(0, fi.Name.Length - 1);
+
+                    if (fi.DeclaringType?.GetProperty(name) != null)
+                        return name;
```
Comment (robertcoltheart) [inline FileHelpers/Fields/FieldBase.cs]: Note that this is probably overkill for detecting F# auto-properties, as in theory you could name a variable ending in `@` using unicode notation.

**Comment by robertcoltheart** [comment]
Thanks for merging 👍. Would be good to get this released ASAP please :)

**Comment by mcavigelli** [comment]
> Thanks for merging 👍. Would be good to get this released ASAP please :)

Thank you for your fix. I will try to do that this weekend.

> @mcavigelli I've tidied this up quite a bit and found a way to compile F# directly in C#, so no need for that F#-only project. However I needed to create a new test project that is not strongly-named, as the Fsharp compiler dll is also not strongly named, so I couldn't use it as part of the `FileHelpers.Tests` project. I had basically 2 alternatives:
> 
>     1. Drop strong-naming for `FileHelpers.Tests` and then have to deal with accessing internal members via reflection, instead of using `InternalsVisibleTo`. This would be quite an invasive change to the tests.
> 
>     2. Create a new test project that is not strongly-named and create the tests for F# there.
> 
> If we want to merge these 2 in the future to make it cleaner, we could do that, and just have to deal with reflection-type stuff then.

I prefer compile time checking over smaller code size: given the two options, I prefer option 2 (the one you chose).

I wonder why the F# compiler dll is not strongly named. Would it be possible to sign the F# compiler dll in a pre-build step with our own strong name key? That would allow to merge the test projects and still not be forced to use reflection.

**Comment by robertcoltheart** [comment]
There are no plans from the F# compiler team to sign the assembly, per: https://github.com/fsharp/FSharp.Compiler.Service/pull/640 Not sure if its worth the effort to do pre-build signing to be honest.

Any chance you could look at releasing this PR out to Nuget? It's blocking some work we have right now.

**Comment by mcavigelli** [comment]
On NuGet you can find a beta version. Let me know if the package works for you.

Publishing a final version should be easy.

**Comment by robertcoltheart** [comment]
Hi @mcavigelli, I've tested this as best as I can and everything seems to work as expected. Please go ahead and push out a final version when you can? Thanks.

**Comment by robertcoltheart** [comment]
Bump, can we push 3.4.0 to nuget please? 👍

**Comment by mcavigelli** [comment]
I have published it Wednesday (March 20th) night, it should be available by now.

**Comment by pantonis** [comment]
@robertcoltheart Is this option available in the latest version of FileHelpers? If yes how can we use it?

**Comment by robertcoltheart** [comment]
Yes, this was released in the latest version.

**Comment by pantonis** [comment]
Thanks

**Pull request merged**